### PR TITLE
[TextFields] Add initial text rect logic

### PR DESCRIPTION
**Pull request by andrewoverton**
This PR adds initial textRect layout calculation logic to MDCBaseTextField and updates the snapshots CORRECTLY as LFS pointers.

Part of #6942.

### CODE REVIEW

File: components/TextFields/src/ContainedInputView/MDCBaseTextField.m [hunk 1]
```diff
@@ -105,6 +106,23 @@ - (MDCBaseTextFieldLayout *)calculateLayoutWithTextFieldSize:(CGSize)textFieldSi
                                                      isEditing:self.isEditing];
 }
 
+/**
+To understand this method one must understand that the CGRect UITextField returns from @c
+-textRectForBounds: does not actually represent the CGRect of visible text in UITextField. It
+represents the CGRect of an internal "field editing" class, which has a height that is significantly
+taller than the text (@c font.lineHeight) itself. Providing a height in @c -textRectForBounds: that
+differs from the height determined by the superclass results in an unusable text field. By taking
```
Comment (jverkoey) [inline components/TextFields/src/ContainedInputView/MDCBaseTextField.m]: What does "unusable text field" mean in this context? Please expand on it here in the docs.

File: components/TextFields/src/ContainedInputView/MDCBaseTextField.m [hunk 2]
```diff
@@ -105,6 +106,23 @@ - (MDCBaseTextFieldLayout *)calculateLayoutWithTextFieldSize:(CGSize)textFieldSi
...
+the desired CGRect of the visible text from the layout object, giving it the height preferred by the
+superclass's implementation of @c -textRectForBounds:, and then ensuring that this new CGRect has
+the same midY as the original CGRect, we are able to take control of the text's positioning.
+ */
+- (CGRect)adjustTextAreaFrame:(CGRect)textRect
+    withParentClassTextAreaFrame:(CGRect)parentClassTextAreaFrame {
+  CGFloat systemDefinedHeight = CGRectGetHeight(parentClassTextAreaFrame);
+  CGFloat minY = CGRectGetMidY(textRect) - (systemDefinedHeight * (CGFloat)0.5);
+  return CGRectMake(CGRectGetMinX(textRect), minY, CGRectGetWidth(textRect), systemDefinedHeight);
+}
```
Comment (jverkoey) [inline components/TextFields/src/ContainedInputView/MDCBaseTextField.m:125]: This sounds like it definitely warrants some unit tests; please add some :)

File: components/TextFields/src/ContainedInputView/private/MDCBaseTextFieldLayout.m [hunk 3]
```diff
@@ -117,4 +140,8 @@ - (BOOL)displaysSideView:(UIView *)subview
   return displaysSideView;
 }
 
+- (CGFloat)textHeightWithFont:(UIFont *)font {
+  return (CGFloat)ceil((double)font.lineHeight);
+}
+
```
Comment (jverkoey) [inline components/TextFields/src/ContainedInputView/private/MDCBaseTextFieldLayout.m:146]: I just noticed that no unit tests were added for this layout logic added in the last PR; please add unit tests for the layout logic (this is a great object to be tested independently).

**Review by jverkoey — CHANGES_REQUESTED** [review]
First pass of feedback broadly requesting more tests to be added for the functionality being modified here.

File: components/TextFields/src/ContainedInputView/MDCBaseTextField.m (hunk 2 above)
Comment (andrewoverton) [reply components/TextFields/src/ContainedInputView/MDCBaseTextField.m:125]: Good call

File: components/TextFields/src/ContainedInputView/private/MDCBaseTextFieldLayout.m (hunk 3 above)
Comment (andrewoverton) [reply components/TextFields/src/ContainedInputView/private/MDCBaseTextFieldLayout.m:146]: I added some initial unit tests for the layout object. I can add more as time goes by and things begin to settle 👍

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldLayoutTests.m [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+@implementation MDCBaseTextFieldLayoutTests
+
+- (void)testTextHeightWithFont {
+  MDCBaseTextFieldLayout *emptyLayout = [[MDCBaseTextFieldLayout alloc] init];
+  UIFont *systemFont = [UIFont systemFontOfSize:[UIFont systemFontSize]];
+  CGFloat ceiledLineHeight = (CGFloat)ceil((double)systemFont.lineHeight);
+  CGFloat calculatedTextHeight = [emptyLayout textHeightWithFont:systemFont];
+  XCTAssertEqual(ceiledLineHeight, calculatedTextHeight);
+}
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldLayoutTests.m:66]: Please test the full set of functionality you've added thus far. Tests will help ensure that we're not regressing on any of the edge cases you've been calling out throughout the code documentation, and it'll be much easier to start writing the tests now than to do them all at once later.

At minimum there needs to be tests for:

- `textRectForBounds:` and the edge case behavior you documented. Specifically write a test for the case you described in the comments so that we don't accidentally regress on it.
- Each of the different configurations currently possible in the layout object (e.g. only a left view, only a right view, both views, no views).
- Each of the different combinations of display modes for both the left and right view.

**Review by jverkoey — CHANGES_REQUESTED** [review]
Please write more comprehensive tests for the functionality you're adding.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldLayoutTests.m (hunk 4 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldLayoutTests.m:66]: I added a bunch of tests for the leading/trailing view stuff.

`-textRectForBounds:` just calls `-adjustTextAreaFrame:withParentClassTextAreaFrame:`, which now has a unit test. Does `-textRectForBounds:` really need to be tested in light of that? Also, I wouldn't call the text rect behavior an edge case, it's just how the UITextField text rect stuff works, in every case.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m [hunk 5]
```diff
@@ -99,4 +101,255 @@ - (void)testTrailingViewModeRTL {
...
+
+- (void)testNoLeadingOrTrailingViewWithViewModeUnlessEditing {
+  MDCBaseTextField *textField =
+      [self createTextFieldWithLeadingView:NO
+                              trailingView:NO
+                                  viewMode:UITextFieldViewModeUnlessEditing];
+  [self forceLayoutOfView:textField];
+  XCTAssertNil(textField.leadingView);
+  XCTAssertNil(textField.trailingView);
+}
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m:169]: Please update the tests to follow our convention of "given/when/then" statements.

Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m:169]: Done.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m [hunk 6]
```diff
@@ -32,71 +34,457 @@ - (UIView *)createSideView {
 }
 
 - (void)testLeadingViewLTR {
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: For each of these tests please consider describing what the test is testing to help with readability.

E.g. `testLeadingViewEqualsLeftViewInLTR`

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m [hunk 7]
```diff
@@ -32,71 +34,457 @@ - (UIView *)createSideView {
...
+  CGRect textAreaFrame = CGRectMake(30, 50, 120, 20);
+  CGRect parentClassTextAreaFrame = CGRectMake(20, 30, 120, 50);
+
+  // When
+  CGRect finalTextAreaFrame =
+      [[[MDCBaseTextField alloc] init] adjustTextAreaFrame:textAreaFrame
+                              withParentClassTextAreaFrame:parentClassTextAreaFrame];
+
+  // Then
+  CGRect correctFinalTextAreaFrame = CGRectMake(30, 35, 120, 50);
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: How were these values chosen? Please either document or, preferably, compute them here so that if the test fails it's clear where these numbers came from.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m [hunk 8]
```diff
@@ -32,71 +34,457 @@ - (UIView *)createSideView {
...
+- (void)forceLayoutOfView:(UIView *)view {
+  [view setNeedsLayout];
+  [view layoutIfNeeded];
+}
+
+- (BOOL)CGRect:(CGRect)CGRect1 isEqualToCGRect:(CGRect)CGRect2 {
+  NSString *CGRect1Description = NSStringFromCGRect(CGRect1);
+  NSString *CGRect2Description = NSStringFromCGRect(CGRect2);
+  return [CGRect1Description isEqualToString:CGRect2Description];
+}
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Please move all helper methods to the bottom or top of the file.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m [hunk 9]
```diff
@@ -32,71 +34,457 @@ - (UIView *)createSideView {
...
+  [window addSubview:textField];
+  return textField;
+}
+
+- (void)forceLayoutOfView:(UIView *)view {
+  [view setNeedsLayout];
+  [view layoutIfNeeded];
+}
+
+- (BOOL)CGRect:(CGRect)CGRect1 isEqualToCGRect:(CGRect)CGRect2 {
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Did you find that `CGRectEqualToRect` wasn't working? If so, please document why here.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m [hunk 10]
```diff
@@ -32,71 +34,457 @@ - (UIView *)createSideView {
...
+                                                               viewMode:UITextFieldViewModeAlways];
+  CGRect leadingFrameNotEditing = textField.leftView.frame;
+
+  // When
+  [textField becomeFirstResponder];
+  CGRect leadingFrameEditing = textField.leftView.frame;
+
+  // Then
+  XCTAssert([self CGRect:leadingFrameNotEditing isEqualToCGRect:leadingFrameEditing]);
+  XCTAssertFalse(textField.leadingView.hidden);
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: tip: to aid with readability here and throughout by removing double-negatives, consider changing this to:

```objc
BOOL leadingViewVisible = !textField.leadingView.hidden;
XCTAssertTrue(leadingViewVisible);
```

Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Does this also need a test for trailingView's visibility?

**Review by jverkoey — CHANGES_REQUESTED** [review]

**Comment by andrewoverton** [comment]
Oof for some reason the tests are passing locally and failing internally >_<

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 9 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: I forgot it existed ahaha

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 10 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: I don't think it makes much of a difference but alright

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 6 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Done

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 7 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Done

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 8 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Done

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 10 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Done

Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: This snippet is outdated now so I'm not positive which test it is but I think it's fine. What do you think?

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldLayoutTests.m (hunk 4 above)
Comment (jverkoey) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldLayoutTests.m:66]: Yes, please test all of the public APIs that you're customizing to ensure that the wiring you've implemented here does not regress.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m [hunk 11]
```diff
@@ -14,89 +14,527 @@
...
+  textField.layoutDirection = UIUserInterfaceLayoutDirectionLeftToRight;
+  if (leadingView) {
+    textField.leadingView = [self createSideView];
+  }
+  if (trailingView) {
+    textField.trailingView = [self createSideView];
+  }
+  textField.trailingViewMode = viewMode;
+  textField.leadingViewMode = viewMode;
+  UIWindow *window = [[UIWindow alloc] init];
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Is it possible to return a text field that has not been added to a UIWindow? If not, please document why we need the UIWindow here.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m [hunk 12]
```diff
@@ -14,89 +14,527 @@
...
 }
 
+- (void)testAdjustTextAreaFrameWithParentClassTextAreaFrame {
+  // Given
+  CGRect desiredTextAreaFrame = CGRectMake(30, 50, 120, 20);
+  CGRect pretendParentClassTextAreaFrame = CGRectMake(20, 30, 120, 50);
+  CGFloat pretendSystemDefinedHeight = CGRectGetHeight(pretendParentClassTextAreaFrame);
+  CGFloat desiredTextAreaMidY = CGRectGetMidY(desiredTextAreaFrame);
+  CGFloat halfOfPretendSystemDefinedHeight = (pretendSystemDefinedHeight * (CGFloat)0.5);
+  CGFloat desiredTextAreaMinY = desiredTextAreaMidY - halfOfPretendSystemDefinedHeight;
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m:156]: tip: generally speaking you don't need to define a variable for each mathematical expression. E.g. consider simply using `(pretendSystemDefinedHeight * (CGFloat)0.5)` here instead.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m [hunk 13]
```diff
@@ -14,89 +14,527 @@
...
+- (BOOL)isEditingOverride {
+  return [objc_getAssociatedObject(self, _cmd) boolValue];
+}
+- (void)setIsEditingOverride:(BOOL)isEditingOverride {
+  objc_setAssociatedObject(self, @selector(isEditingOverride), @(isEditingOverride),
+                           OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+- (BOOL)isEditing {
+  return self.isEditingOverride ? self.isEditingOverride : [super isEditing];
+}
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Overriding an existing method in a class can potentially lead to flaky tests because the order in which overloaded category methods are loaded is undefined.

Did you find that becoming the first responder was not sufficient within the test harness? If so then either a private test-only API or a subclass are two alternative solutions here.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m [hunk 14]
```diff
@@ -14,89 +14,527 @@
...
+  CGRect leadingFrameEditing = textField.leftView.frame;
+  CGRect trailingFrameEditing = textField.rightView.frame;
+
+  // Then
+  XCTAssertFalse(leadingViewVisibleBeforeBecomingFirstResponder);
+  XCTAssertFalse(trailingViewVisibleBeforeBecomingFirstResponder);
+  XCTAssertTrue(leadingViewVisibleAfterBecomingFirstResponder);
+  XCTAssertTrue(trailingViewVisibleAfterBecomingFirstResponder);
+  XCTAssertFalse(CGRectEqualToRect(leadingFrameNotEditing, leadingFrameEditing));
+  XCTAssertFalse(CGRectEqualToRect(trailingFrameNotEditing, trailingFrameEditing));
```
Comment (jverkoey) [inline components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: General question: is it generally true that the leading/trailing frames will change when visibility changes? I'd expect only the views' hidden property to be toggled; is it possible for us to make the text field do so? Adjusting the frame of the left/right views could cause costly operations to occur (e.g. redrawing an icon), so if it's possible to easily avoid excessive layouts then that would be preferable.

**Review by jverkoey — APPROVED** [review]
Generally LGTM but please address the remaining comments.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 14 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Not sure how it works with UITextField. I should look. With mine the way it's worked so far is their origins are made to be {0, 0}, but the size doesn't change, and they're hidden. This isn't because I felt it was the best way to do it necessarily, it was just one way that seemed to work so didn't think about it anymore.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 13 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Yeah this is sketchy, I'm just trying a bunch of random stuff right now to get the tests to pass in kokoro. Having a hard time!

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 11 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: None of the tests passed locally until I added them to a window :/ . Still can't get them to pass on kokoro, but that's another story.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 14 above)
Comment (jverkoey) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Ah; can we simplify it to just toggle the hidden property then?

Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: I only bother calculating origins of the side views if they're going to be displayed--that's why they're { 0, 0 } otherwise. It's actually simpler if I keep it the way it is. Is that alright?

**Comment by andrewoverton** [comment]
@jverkoey , unit testing the leadingView/trailingView/leadingViewMode/trailingViewMode stuff is really hard--I'm afraid there might not be a way to get UITextField to work the way it does in an app from a unit test that doesn't have an app host. However, there seems to be a way with snapshot tests. Adding the textfield to the app host's key window and waiting ~1 second seems to do the trick. How would you feel about getting rid of all these leading/trailing unit tests I've written and replacing them with snapshot tests that make use of this strategy? See the ["testEditingTextFieldWithLeadingViewWhileEditing"](https://github.com/material-components/material-components-ios/blob/31904b61c89c2bd71bca623f3da514ab1887c05f/snapshot_test_goldens/goldens_64/MDCBaseTextFieldTestsSnapshotTests/testEditingTextFieldWithLeadingViewWhileEditing_11_2%402x.png) and ["testNonEditingTextFieldWithLeadingViewWhileEditing"](https://github.com/material-components/material-components-ios/blob/31904b61c89c2bd71bca623f3da514ab1887c05f/snapshot_test_goldens/goldens_64/MDCBaseTextFieldTestsSnapshotTests/testNonEditingTextFieldWithLeadingViewWhileEditing_11_2%402x.png) snapshots for proofs of concept. If you're okay with this general idea I'll make them nicer.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 14 above)
Comment (jverkoey) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Extra paths of logic add complexity to a code base, and if you can accomplish the same thing with less code then that's usually preferable. I'm not seeing a strong case for why we need to conditionally update the layout of the views based on their visibility, so in this case the simpler logical path will be to simply update the visibility.

Please do so, barring a strong technical reason why it's beneficial to go with your current approach.

**Comment by jverkoey** [comment]
> unit testing the leadingView/trailingView/leadingViewMode/trailingViewMode stuff is really hard--I'm afraid there might not be a way to get UITextField to work the way it does in an app from a unit test that doesn't have an app host.

Did you evaluate testing just an instance of the layout object?

I see that you've mostly been adding full integration tests of the component, which will require a greater deal of coordination as you're encountering. Snapshot tests are a great way to validate the overall functionality of a component, so you can to some degree rely on them to achieve the level of integration testing you're trying to do here, yes.

With that in mind, one of the benefits of you having created a layout object is that you can succinctly test only that object as a single unit. This is what would be traditionally considered a *unit test*. You'll find that this type of testing enables you to test the different layout combinations the component is capable of without the constraints of a host environment.

Put in specific actionable terms:

- Move any tests in MDCBaseTextFieldTests.m that have to do specifically with layout to MDCBaseTextFieldLayoutTests.m instead.
- All tests within MDCBaseTextFieldLayoutTests.m should only ever work with an instance of the layout object.

**Comment by andrewoverton** [comment]
@jverkoey , Ah yep, I thought you wanted me to test the leading/trailing behavior on the textfield, I didn't realize the layout would be sufficient. I can port these troublesome leading/trailing unit tests to just test the layout object, that will be much easier. Still, I think figuring out how to get the editing behavior working for snapshot tests is useful. That will probably come into play later.

File: components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m (hunk 14 above)
Comment (andrewoverton) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Here's some pseudocode that shows how the left view frame is calculated:
```
if (shouldShowLeftView) {
    // Do the work of calculating what the origin should be so the frame looks okay
} else {
    // Don't bother, because nobody will see it
} 
```
If you want these invisible views to have certain frames at the cost of a little more math i'm more than happy to do that for you

Comment (jverkoey) [reply components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m]: Yes, please remove that conditional path. We can add it or something like it back in the future if we determine that it adversely affects performance.
